File: 0017 - Letter Combinations of a Phone Number/Letter Combinations of a Phone Number.cpp
const int ZERO = [](){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    return 0;
}();

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> result;
        vector<string> numbers = {"0", "1", "abc", "def", "ghi", "jkl", "mno",
                                 "pqrs", "tuv", "wxyz"};
        if (digits.size() == 0) {
            return {};
        }
        backtrack(numbers, result, "", digits, 0);
        return result;
    }
    
    void backtrack(vector<string>& numbers, vector<string>& result, string path, string digits, int index) {
        if (path.size() == digits.size()) {
            result.push_back(path);
            return;
        }
        
        string letter = numbers[digits[index] - '0'];
        for (int i = 0; i < letter.size(); ++i) {
            backtrack(numbers, result, path + letter[i], digits, index + 1);
        }
    }
};

File: 0017 - Letter Combinations of a Phone Number/Letter Combinations of a Phone Number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0:
            return []
        
        numbers = ["0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
        result = []
        
        self.backtrack(result, digits, numbers, '', 0)
        
        return result
        
    def backtrack(self, result: List[str], digits: str, numbers: List[str], path: str, index: int) -> None:
        if len(path) == len(digits):
            result.append(path)
            return
        letters = numbers[int(digits[index])]
        
        for i in range(len(letters)):
            self.backtrack(result, digits, numbers, path + letters[i], index + 1)
